View all followers

0nko opened the pull request:
This is an intermediate PR that only adds a new View all screen for the followers (too keep the PRs small).

To test:
1. Go to Stats
2. Find the Followers block and tap View more
3. Notice a new activity is opened and 10 items are immediately loaded
4. Tap on the Load more button at the bottom
5. Notice 10 more items are loaded
6. Keep tapping the button until all available items are loaded
7. After everything is loaded on the screen, the Load more button is gone

![image](https://user-images.githubusercontent.com/1522856/52418059-8c3e0000-2aed-11e9-9e2c-482ad09be5a1.png)

0nko committed:
Add an activity, fragment, view model and layout for the View all screen

0nko committed:
Merge branch 'develop' into issue/8815-view-more-screen

0nko committed:
Remove unused code

0nko committed:
Update the InisightsStore package references

0nko committed:
Merge branch 'develop' into issue/8815-view-more-screen

0nko committed:
Refactor the StatsViewAllViewModel

0nko committed:
Reference the latest FluxC hash

0nko committed:
Add the View all activity to the manifest

0nko committed:
Add the ViewModel, fragment and use followers use case view all screen provider for Dagger

0nko committed:
Add the use case loading logic to the View all ViewModel

0nko committed:
Pass the required arguments to the View all screen and initialize the fragment

0nko committed:
Point to the latest relevant FluxC hash

0nko committed:
Update the View all stats activity title

0nko committed:
Rename the UseCaseFactory to GranularUseCaseFactory to distinguish it from Insights

0nko committed:
Add a toolbar to the View all screen layout

0nko committed:
Create FollowersUseCase using a new factory and use paging when fetching in the View all mode

0nko committed:
Merge branch 'issue/8815-view-more-screen' into issue/view-more-followers

0nko committed:
Remove unused arguments

0nko committed:
Optimize imports

0nko committed:
Merge branch 'develop' into issue/view-more-followers

0nko committed:
Fix the failing unit tests

0nko committed:
Add tests for loading followers with paging

0nko committed:
Merge branch 'issue/8815-view-more-screen' into issue/view-more-followers

planarvoid commented:
The app crashes on rotation with:
```
Caused by: java.lang.ClassCastException: org.wordpress.android.ui.stats.StatsViewType cannot be cast to org.wordpress.android.fluxc.model.SiteModel
        at org.wordpress.android.ui.stats.refresh.StatsViewAllFragment.initializeViewModels(StatsViewAllFragment.kt:151)
        at org.wordpress.android.ui.stats.refresh.StatsViewAllFragment.onViewCreated(StatsViewAllFragment.kt:136)
```

planarvoid commented:
Overall this looks pretty good :+1: 
I have few things (that might be addressed in another PR):
* In the landscape we are showing the data with the staggered grid layout manager which shows only 1 column - I think we should just use the normal LinearLayoutManager.
* We show the `Load more` button when there is no data in the list (for example when the `E-MAIL` followers are empty)
* Are we planning to implement infinite paging? trigger next page load on scroll?

planarvoid commented on WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/insights/usecases/FollowersUseCase.kt:
```diff
@@ -25,36 +26,52 @@ import org.wordpress.android.ui.stats.refresh.lists.sections.BlockListItem.ListI
-        val wpComFollowers = insightsStore.getWpComFollowers(site, PAGE_SIZE)
-        val emailFollowers = insightsStore.getEmailFollowers(site, PAGE_SIZE)
+        lastSite = site
+        val wpComFollowers = insightsStore.getWpComFollowers(site, pageSize)
```
I'm wondering if this will always work as expected. I think we should load all the cached data when we're opening the `ViewAll` screen. What about making the pageSize nullable and pass null when we want all the items in the DB? I think it makes sense to load everything we have.

0nko commented on WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/StatsViewAllFragment.kt:
```diff
@@ -0,0 +1,335 @@
+    }
+}
+
+sealed class NavigationTarget : Event() {
```
Agreed. I just forgot to delete it when I copied the fragment file.

planarvoid commented on WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/insights/usecases/FollowersUseCase.kt:
```diff
@@ -25,36 +26,52 @@ import org.wordpress.android.ui.stats.refresh.lists.sections.BlockListItem.ListI
     private val resourceProvider: ResourceProvider,
-    private val analyticsTracker: AnalyticsTrackerWrapper
+    private val analyticsTracker: AnalyticsTrackerWrapper,
+    private val isViewAllMode: Boolean
```
I think it might be a bit more readable to have this as an enum `UseCaseMode` with values `CARD, VIEW_ALL`.

0nko committed:
Add the loadMode argument to fetch methods in all stats use cases

0nko committed:
Update FluxC reference

0nko committed:
Merge branch 'issue/8815-view-more-screen' into issue/view-more-followers

0nko committed:
Merge branch 'issue/8815-view-more-screen' into issue/view-more-followers

# Conflicts:
#	WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/StatsModule.kt
#	WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/granular/usecases/DateUseCase.kt
#	build.gradle

0nko committed:
Propagate the loadMode parameter up to the BaseListUseCase

0nko committed:
Fix the incorrect type deserialization

0nko committed:
Remove unused class

0nko committed:
Move the NavigationTarget class to a separate file

0nko committed:
Moved the navigation logic to a separate class StatsNavigator

0nko committed:
Remove unused code

0nko committed:
Use the common navigator class in the navigation target observer

0nko committed:
Remove unused code

0nko committed:
Move an extension function to StatsNavigator

0nko committed:
Optimize imports

0nko committed:
Use an explicit enum value to signify the view-all mode in use cases

0nko committed:
Update the FluxC reference

0nko committed:
Merge branch 'issue/8815-view-more-screen' into issue/view-more-followers

0nko committed:
Remove the unnecessary loadMode parameter from the use case layer

0nko committed:
Update the unit tests

0nko committed:
Update the View all followers loading code with the changes in FluxC

0nko committed:
Fix a typo

0nko committed:
Always use a linear layout manager in the View all screen

0nko committed:
Add a base View all ViewModel

0nko committed:
Add a custom view all layout that has pull-to-refresh

0nko committed:
Implement pull-to-refresh in the view all screen

0nko committed:
Force the data to reload when force = true

0nko committed:
Optimize imports

0nko commented:
Thanks for a proper review, Vojta. I think I've addressed all of the comments. 

- I added a common StateNavigator
- I added a way to request how the data is loaded from the cach (`CacheMode`) and the server (`FetchMode`)
- Any data already downloaded is now shown when the View all scren opens
- I added a pull-to-refresh
- I added some unit tests
- Rotation crash is fixed and the already-loaded data is just displayed after rotation
- I changed the layout to LineaLayoutManager

> We show the `Load more` button when there is no data in the list (for example when the `E-MAIL` followers are empty)

This was the existing behavior in the block too.

> Are we planning to implement infinite paging? trigger next page load on scroll?

Sure, but let's do it in a different PR, this one's already huge.

0nko committed:
Update the FluxC reference

0nko committed:
Fix the duplicate entry issue

0nko committed:
Fix the failing unit test

0nko committed:
Merge branch 'issue/8815-view-more-screen' into issue/view-more-followers

planarvoid commented on WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/StatsViewAllFragment.kt:
```diff
@@ -0,0 +1,172 @@
+
+        val clazz = when (type) {
+            StatsViewType.FOLLOWERS -> StatsViewAllFollowersViewModel::class.java
+            else -> StatsViewAllFollowersViewModel::class.java
```
NIT: I'd replace this with a TODO() and probably list all the branches

planarvoid commented on WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/utils/StatsNavigator.kt:
```diff
@@ -0,0 +1,160 @@
+import org.wordpress.android.ui.stats.refresh.NavigationTarget.ViewVideoPlays
+import org.wordpress.android.util.ToastUtils
+
+class StatsNavigator(
```
I think it would be better if this class were injected and the `navigate` method had `site` and `activity` as parameters. Especially since it's used in 2 fragments.

planarvoid reviewed (approved):
this looks awesome! 💯  consider my 2 proposals but otherwise feel free to merge it and the FluxC PR

0nko committed:
Add TODO stubs for future view-all screen types

0nko committed:
Inject the StatsNavigator instead of direct instantiation
